bindings/aws/s3/s3.go

```diff
@@ -44,39 +71,168 @@ func (s *AWSS3) Init(metadata bindings.Metadata) error {
+		Bucket: aws.String(metadata.Bucket),
 		Key:    aws.String(key),
 		Body:   r,
 	})
+	if err != nil {
+		return nil, fmt.Errorf("s3 binding error. Uploading: %w", err)
+	}
+
+	jsonResponse, err := json.Marshal(createResponse{
+		Location:  resultUpload.Location,
+		VersionID: resultUpload.VersionID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("s3 binding error. Error marshalling create response: %w", err)
+	}
+
+	return &bindings.InvokeResponse{
+		Data: jsonResponse,
+	}, nil
+}
+
+func (s *AWSS3) get(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	var key string
+	if val, ok := req.Metadata[metadataKey]; ok && val != "" {
+		key = val
+	} else {
+		return nil, fmt.Errorf("s3 binding error: can't read key value")
+	}
+
+	buff := &aws.WriteAtBuffer{}
+
+	_, err := s.downloader.Download(buff,
+		&s3.GetObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(key),
+		})
+	if err != nil {
+		return nil, fmt.Errorf("s3 binding error: error downloading S3 object: %w", err)
+	}
+
+	return &bindings.InvokeResponse{
+		Data:     buff.Bytes(),
+		Metadata: nil,
+	}, nil
+}
+
+func (s *AWSS3) delete(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	var key string
+	if val, ok := req.Metadata[metadataKey]; ok && val != "" {
+		key = val
+	} else {
+		return nil, fmt.Errorf("s3 binding error: can't read key value")
+	}
+
+	_, err := s.s3Client.DeleteObject(
+		&s3.DeleteObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(key),
+		})
 
 	return nil, err
 }
 
+func (s *AWSS3) list(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	var payload listPayload
+	err := json.Unmarshal(req.Data, &payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if payload.MaxResults == int32(0) {
+		payload.MaxResults = maxResults
+	}
+
+	input := &s3.ListObjectsInput{
+		Bucket:    aws.String(s.metadata.Bucket),
+		MaxKeys:   aws.Int64(int64(payload.MaxResults)),
+		Marker:    aws.String(payload.Marker),
+		Prefix:    aws.String(payload.Prefix),
+		Delimiter: aws.String(payload.Delimiter),
+	}
+
+	result, err := s.s3Client.ListObjects(input)
+	if err != nil {
+		return nil, fmt.Errorf("s3 binding error. list operation. cannot marshal blobs to json: %w", err)
+	}
+
+	jsonResponse, err := json.Marshal(result)
+	if err != nil {
+		return nil, fmt.Errorf("s3 binding error. list operation. cannot marshal blobs to json: %w", err)
+	}
+
+	return &bindings.InvokeResponse{
+		Data: jsonResponse,
+	}, nil
+}
+
+func (s *AWSS3) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	switch req.Operation {
+	case bindings.CreateOperation:
+		return s.create(req)
+	case bindings.GetOperation:
+		return s.get(req)
+	case bindings.DeleteOperation:
+		return s.delete(req)
+	case bindings.ListOperation:
+		return s.list(req)
+	default:
+		return nil, fmt.Errorf("s3 binding error. unsupported operation %s", req.Operation)
```

Comment:
Please, default to CreateOperation for backwards compatibility.

Reply:
I think it isn’t necessary.

When you invoke the binding, operation info is required. If you invoke the binding and the operation isn’t defined in bindings.OperationKind method you will get an error.

If you add a new operation in bindings.OperationKind method and you don’t implement it in Invoke method I think it’s better to get an error instead of execute Create operation

Reply:
This component is still in Alpha, which means breaking changes are allowed. Making it backwards compatible is not necessary but we still avoid breaking existing users unnecessarily.

To favor existing customer experience, I think we should make this backwards compatible. It is not a big change.

/cc @pkedy @CodeMonkeyLeet thoughts?

Reply:
Yes, it's an easy change but not needed.  The changes are compatible and we aren't going to break any application:


All request must include operation field if you don't include it you will get an error also you will get an error if you try invoke and operation not defined in bindings.OperationKind method.

![image](https://user-images.githubusercontent.com/5806680/132370100-a02690d5-2b67-480a-805b-9eab73947631.png)

Reply:
OK. I got it. It is being enforced prior to the component code.